cmd/terminal.go

```diff
@@ -0,0 +1,367 @@
+	ticker := time.NewTicker(pingPeriod)
+	defer ticker.Stop()
+
+	conn.SetPongHandler(func(string) error {
+		ticker.Reset(pingPeriod)
+		return conn.SetReadDeadline(time.Now().Add(pongWait))
+	})
+
+	conn.SetPingHandler(func(msg string) error {
+		ticker.Reset(pingPeriod)
+		err := conn.SetReadDeadline(time.Now().Add(pongWait))
+		if err != nil {
+			return err
+		}
+		return conn.WriteControl(
+			websocket.PongMessage,
+			[]byte(msg),
+			time.Now().Add(writeWait),
+		)
+	})
+
+	log.Info("Press CTRL+] to quit the session")
+
+	// set the terminal in raw mode
+	oldState, err := term.MakeRaw(termID)
+	if err != nil {
+		return errors.Wrap(err, "Unable to set the terminal in raw mode")
+	}
+	defer func() {
+		_ = term.Restore(termID, oldState)
+	}()
+
+	// send the shell start message
+	m := &ws.ProtoMsg{
+		Header: ws.ProtoHdr{
+			Proto:   ws.ProtoTypeShell,
+			MsgType: wsshell.MessageTypeSpawnShell,
+			Properties: map[string]interface{}{
+				"terminal_width":  termWidth,
+				"terminal_height": termHeight,
+			},
+		},
+	}
+
+	data, err := msgpack.Marshal(m)
+	if err != nil {
+		return errors.Wrap(err, "Unable to parse the message from the websocket")
+	}
+	_ = conn.SetWriteDeadline(time.Now().Add(writeWait))
+	_ = conn.WriteMessage(websocket.BinaryMessage, data)
+
+	// message channel
+	msgChan := make(chan *ws.ProtoMsg)
+
+	c.running = true
+	go c.pipeStdout(msgChan, os.Stdout)
+	go c.pipeStdin(conn, os.Stdin)
+
+	// handle CTRL+C and signals
+	quit := make(chan os.Signal, 1)
+	signal.Notify(quit, unix.SIGINT, unix.SIGTERM)
+
+	// resize the terminal window
+	go c.resizeTerminal(ctx, msgChan, termID, termWidth, termHeight)
+
+	// wait for CTRL+C, signals or stop
+	for c.running {
+		select {
+		case msg := <-msgChan:
+			data, err := msgpack.Marshal(msg)
+			if err != nil {
+				log.Err(fmt.Sprintf("error: %v", err))
+				break
+			}
+			_ = conn.SetWriteDeadline(time.Now().Add(writeWait))
+			_ = conn.WriteMessage(websocket.BinaryMessage, data)
+		case <-ticker.C:
+			pongWaitString := strconv.Itoa(int(pongWait.Seconds()))
+			_ = conn.WriteControl(
+				websocket.PingMessage,
+				[]byte(pongWaitString),
+				time.Now().Add(writeWait),
+			)
+		case <-quit:
+			c.running = false
+		case <-c.stop:
+			c.running = false
+		}
+	}
+
+	// cancel the context
+	cancelContext()
+
+	// send the stop shell message
+	m = &ws.ProtoMsg{
+		Header: ws.ProtoHdr{
+			Proto:     ws.ProtoTypeShell,
+			MsgType:   wsshell.MessageTypeStopShell,
+			SessionID: c.sessionID,
+		},
+	}
+
+	data, err = msgpack.Marshal(m)
+	if err != nil {
+		return errors.Wrap(err, "Unable to parse the message from the websocket")
+	}
+	_ = conn.SetWriteDeadline(time.Now().Add(writeWait))
+	_ = conn.WriteMessage(websocket.BinaryMessage, data)
+	conn.Close()
+
+	return c.err
+}
+
+func (c *TerminalCmd) resizeTerminal(ctx context.Context, msgChan chan *ws.ProtoMsg, termID int, termWidth int, termHeight int) {
+	ticker := time.NewTicker(500 * time.Millisecond)
```

Comment:
Isn't this a bit excessive? Can't we use the SIGWINCH signal to detect terminal resize?

Reply:
+1

Reply:
I'm sorry, I didn't think about SIGWINCH, thanks for the suggestion!